src/plugins/discourse/mirror.js

```diff
@@ -67,15 +73,109 @@ export class Mirror {
       ...defaultOptions,
       ...(options || {}),
     };
+    this._mode = mode || 1;
   }
 
   async update(reporter: TaskReporter) {
     reporter.start("discourse");
-    await this._updateTopics(reporter);
-    await this._updateLikes(reporter);
+    switch (this._mode) {
+      case 1:
+        await this._updateTopics(reporter);
+        await this._updateLikes(reporter);
+        break;
+      case 2:
+        await this._updateTopicsV2(reporter);
+        await this._updateLikes(reporter);
+        break;
+      default:
+        throw new Error(`Invalid mode ${this._mode}, expecting 1 or 2`);
+    }
     reporter.finish("discourse");
   }
 
+  async _updateTopicsV2(reporter: TaskReporter) {
+    reporter.start("discourse/topics");
+
+    const {
+      topicBumpMs: lastLocalTopicBumpMs,
+      definitionCheckMs: lastDefinitionCheckMs,
+    } = this._repo.syncHeads();
+
+    const startTime = Date.now();
+    const shouldCheckDefinitions =
+      startTime - lastDefinitionCheckMs >=
+      this._options.recheckCategoryDefinitionsAfterMs;
+
+    const bumpedTopics: TopicLatest[] = await this._fetcher.topicsBumpedSince(
+      lastLocalTopicBumpMs
+    );
+    const topicBumpsById: Map<number, number> = new Map(
+      bumpedTopics.map((t) => [t.id, t.bumpedMs])
+    );
+
+    // Make sure we have oldest first in our load queue.
+    bumpedTopics.sort(sortTopicByBumpedMs);
+
+    // Create a uniqueness filter.
+    const existingTopics = new Set();
+    const once = (tid) => {
+      if (existingTopics.has(tid)) return false;
+      existingTopics.add(tid);
+      return true;
+    };
+
+    // Add definition topics if the flag to do so is set.
+    const definitionTopicIds = shouldCheckDefinitions
+      ? await this._fetcher.categoryDefinitionTopicIds()
+      : [];
+
+    // Add specific-category topics when it's set in our options.
+    const selectCategoryTopics = this._repo.topicsInCategories(
+      this._options.recheckTopicsInCategories
+    );
+
+    // Note: order is important here.
+    // Initial load should happen in order of bump date,
+    // reloads at the end are ok to be in random order.
+    const topicLoadQueue = [
+      ...bumpedTopics.map((t) => t.id),
+      ...definitionTopicIds,
+      ...selectCategoryTopics,
+    ].filter(once);
+
+    for (const topicId of topicLoadQueue) {
+      const topicWithPosts = await this._fetcher.topicWithPosts(topicId);
+      if (topicWithPosts != null) {
```

Comment:
The fetcher returns a null when the API returns a 403, 404 or 410 status.
This typically happens when topics are deleted or made private.
(In mode 1, it would also happen because we were ID guessing)

We might consider deleting the topic from the mirror when this happens, but as this wouldn't be enough to implement topic deletion correctly, we're ignoring it instead.

As you can never really exclude a race condition, I don't think we should warn or crash here.
For example, one might assume a topic with this ID must exist, because we found it in the list of recently bumped topics. But if we have 200 topics to fetch that still takes several minutes to do and topics can be deleted in that time.
More likely though, the `recheckTopicsInCategories` option finds it's topic IDs in the local database, so they could be weeks old and refer to a topic that's been deleted.